kythe/rust/indexer/src/providers.rs

```diff
@@ -0,0 +1,152 @@
+  /// the file exists and that the contents can be requested.
+  ///
+  /// # Errors
+  ///
+  /// If the file is not a valid zip archive, a [KytheError::KzipFileError] will
+  /// be returned.
+  fn exists(&mut self, file_hash: &str) -> bool;
+
+  /// Retrieves the string contents of a file.
+  ///
+  /// This function takes a file hash and returns the string contents of the
+  /// file.
+  ///
+  /// # Errors
+  ///
+  /// If the file does not exist, a
+  /// [FileNotFoundError][KytheError::FileNotFoundError] will be returned.
+  /// If the file can't be read into a String, a
+  /// [FileReadError][KytheError::FileReadError] will be returned.
+  fn contents(&mut self, file_hash: &str) -> Result<String, KytheError>;
+}
+
+/// A [FileProvider] that backed by a .kzip file.
+pub struct KzipFileProvider {
+  file: File,
+  zip_archive: Archive,
+}
+
+impl KzipFileProvider {
+  /// Create a new instance using a valid [File] object of a kzip
+  /// file.KytheError
+  ///
+  /// # Errors
+  ///
+  /// If the file is not a valid zip archive, a
+  /// [KzipFileError][KytheError::KzipFileError] will be returned.
+  pub fn new(file: File) -> Result<Self, KytheError> {
+    let zip_archive = match file.read_zip() {
+      Ok(zip_archive) => zip_archive,
+      Err(_e) => return Err(KytheError::KzipFileError),
+    };
+    Ok(Self { file, zip_archive })
+  }
+
+  /// Retrieve the Compilation Units from the kzip.
+  ///
+  /// This function will attempt to parse the proto files in the `pbunits`
+  /// folder into Compilation Units, returning a Vector of them.
+  ///
+  /// # Errors
+  ///
+  /// If a file cannot be parsed, a
+  /// [ProtobufParseError][KytheError::ProtobufParseError] will be returned.
+  pub fn get_compilation_units(&self) -> Result<Vec<CompilationUnit>, KytheError> {
+    let mut compilation_units = Vec::new();
+    for entry in self.zip_archive.entries() {
+      // Protobuf files are in the pbunits folder
+      if entry.name().contains("/pbunits/") {
+        // Read the file into a compilation unit protobuf
+        match entry.contents() {
+          EntryContents::File(f) => {
+            let mut entry_reader = f
+              .entry
+              .reader(|offset| positioned_io::Cursor::new_pos(&self.file, offset));
+            let bundle = match ::protobuf::parse_from_reader::<CompilationBundle>(&mut entry_reader)
+            {
+              Ok(bundle) => bundle,
+              Err(_e) => return Err(KytheError::ProtobufParseError),
+            };
+            compilation_units.push(bundle.get_unit().clone());
+          }
+          _ => {}
+        };
+      }
+    }
+    Ok(compilation_units)
+  }
+
+  /// Retrieve a file's location in the kzip
+  ///
+  /// This function takes a file hash and returns an Option of the file's index
+  /// in the zip archive's file entries.
+  fn get_file_position(&self, file_hash: &str) -> Option<usize> {
+    self.zip_archive.entries().iter().position(|file_entry| {
```

Comment:
Is this called often? If so,it might be worth precomputing an index in `new`. Otherwise, it might be good to leave a comment justifying the linear search.

Reply:
Good point. I guess it is a tradeoff between time vs space complexity. Would a precomputed HashMap between file hash and index be the best implementation here?

Reply:
That would be a fine approach, if the complexity is justified (i.e. if zip archives might contain many files and/or this function is called very frequently)

Reply:
Implemented the HashMap in the newest commit